### PR TITLE
fix version_info if version can not be determined

### PR DESCRIPTION
**Pull request by esc**
The current implementation assumes that the version info can be split
into at least three parts using the period as delimiter. In certain
situations, such as running from a dowloaded release archive, or when
building from a git checkout that does not have any tags, this
assumption does not hold.

Fixes: #6125

### CODE REVIEW

**Comment by esc** [comment]
I considered some tests for this function, but then was discouraged, as `_gen_v_info` is deleted again after having been defined and used and because the function is in `numba/__init__.py` I am not sure it is worth the effort to mock the `del` in that module during testing. I was unsure what the intent of this is, @stuartarchibald can you clarify? I think I would prefer to keep the function as a utility and additionally add something like `myversion` as a parameter so that it can operate on arbitrary versions rather than always getting the `__version__` from `numba`. That would also significantly ease testing as nothing would need to be mocked.

**Comment by stuartarchibald** [comment]
> I considered some tests for this function, but then was discouraged, as `_gen_v_info` is deleted again after having been defined and used and because the function is in `numba/__init__.py` I am not sure it is worth the effort to mock the `del` in that module during testing. I was unsure what the intent of this is, @stuartarchibald can you clarify?

Didn't want to pollute the top level namespace.

> I think I would prefer to keep the function as a utility and additionally add something like `myversion` as a parameter so that it can operate on arbitrary versions rather than always getting the `__version__` from `numba`. That would also significantly ease testing as nothing would need to be mocked.

Could the function be moved into `numba.misc.init` or something, imported in `__init__`, run and results stored, then deleted. This keeps the top level cleaner and also gives opportunity to test? It'd also present an opportunity to get a load of the other functions out of `__init__` and somewhere testable.

**Comment by esc** [comment]
This has now been updated and is ready for the first round of reviews.

File: numba/__init__.py [hunk 1]
```diff
@@ -6,34 +6,15 @@
 import re
 import sys
 import warnings
-from collections import namedtuple
 
 from ._version import get_versions
+from .misc.init_utils import generate_version_info
```
Comment (stuartarchibald) [inline numba/__init__.py:11]:
```suggestion
from numba.misc.init_utils import generate_version_info
```

File: numba/misc/init_utils.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+from collections import namedtuple
```
Comment (stuartarchibald) [inline numba/misc/init_utils.py:1]: Module docstring?

File: numba/misc/init_utils.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+version_info = namedtuple('version_info',
+                          ('major minor patch short full '
+                           'string tuple git_revision'))
+
+
+def generate_version_info(version):
```
Comment (stuartarchibald) [inline numba/misc/init_utils.py:8]: Docstring?

File: numba/tests/test_inti_utils.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import unittest
```
Comment (stuartarchibald) [inline numba/tests/test_inti_utils.py:1]: This test module has a typo. s/inti/init/

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, just a few minor things to resolve else good to go.

**Comment by stuartarchibald** [comment]
https://github.com/numba/numba/issues/6167 is probably the cause of CI failure.

File: numba/misc/init_utils.py (hunk 2 above)
Comment (esc) [reply numba/misc/init_utils.py:1]: Sure

File: numba/misc/init_utils.py (hunk 3 above)
Comment (esc) [reply numba/misc/init_utils.py:8]: sure

File: numba/tests/test_inti_utils.py (hunk 4 above)
Comment (esc) [reply numba/tests/test_inti_utils.py:1]: indeed

**Comment by esc** [comment]
Changes have been made and the PR is ready for the next round of reviews.

File: numba/misc/init_utils.py [hunk 5]
```diff
@@ -6,6 +8,23 @@
 
 
 def generate_version_info(version):
+    """Process a version string into a structured version_info object.
+
+    Parameters
+    ----------
+    version: str
+        a string describing
```
Comment (stuartarchibald) [inline numba/misc/init_utils.py:16]:
```suggestion
        a string describing the version
```

File: numba/misc/init_utils.py [hunk 6]
```diff
@@ -6,6 +8,23 @@
...
+
+    Returns
+    -------
+    version_info: tuple
+        structured version information
+
+    See also
+    --------
+    Look at the definition of 'version_info' in this file for details.
```
Comment (stuartarchibald) [inline numba/misc/init_utils.py:25]:
```suggestion
    Look at the definition of 'version_info' in this module for details.
```

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fixes.

**Comment by stuartarchibald** [comment]
CI fail is unrelated: https://github.com/numba/numba/issues/5973

**Pull request merged**